File: src/test/java/com/example/petclinic/service/OwnerServiceTest.java
package com.example.petclinic.service;

import com.example.petclinic.domain.entity.Owner;
import com.example.petclinic.domain.repository.OwnerRepository;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import jakarta.validation.Validation;
import jakarta.validation.Validator;
import jakarta.validation.ConstraintViolation;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.Pattern;

import java.util.List;
import java.util.Set;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

class OwnerServiceTest {

    @Mock
    private OwnerRepository ownerRepository;

    @InjectMocks
    private OwnerService ownerService;

    private final Validator validator = Validation.buildDefaultValidatorFactory().getValidator();

    private Owner owner;

    @BeforeEach
    void setUp() {
        MockitoAnnotations.openMocks(this);
        owner = new Owner();
        owner.setFirstName("John");
        owner.setLastName("Doe");
        owner.setAddress("123 Main St");
        owner.setCity("Springfield");
        owner.setTelephone("123456789");
    }

    @Test
    void shouldValidateOwnerRequest_emptyDTO_failsValidation() {
        Owner emptyOwner = new Owner();

        Set<ConstraintViolation<Owner>> violations = validator.validate(emptyOwner);

        assertEquals(4, violations.size());
        assertTrue(violations.stream().anyMatch(v -> v.getPropertyPath().toString().equals("firstName")));
        assertTrue(violations.stream().anyMatch(v -> v.getPropertyPath().toString().equals("lastName")));
    }

    @Test
    void shouldCreateOwner() {
        when(ownerRepository.save(any(Owner.class))).thenAnswer(invocation -> invocation.getArgument(0));

        Owner result = ownerService.save(owner);

        assertNotNull(result);
        assertEquals("John", result.getFirstName());
        assertEquals("Doe", result.getLastName());
        verify(ownerRepository, times(1)).save(any(Owner.class));
    }

    @Test
    void shouldFindAllOwners() {
        when(ownerRepository.findAll()).thenReturn(List.of(owner));

        List<Owner> results = ownerService.findAll();

        assertNotNull(results);
        assertEquals(1, results.size());
        assertEquals("John", results.get(0).getFirstName());
    }
}

File: src/test/java/com/example/petclinic/service/OwnerServiceTest.kt
package com.example.petclinic.service

import com.example.petclinic.domain.entity.Owner
import com.example.petclinic.domain.repository.OwnerRepository
import io.kotest.matchers.collections.shouldContainAll
import io.kotest.matchers.nulls.shouldBeNull
import io.kotest.matchers.nulls.shouldNotBeNull
import io.kotest.matchers.shouldBe
import jakarta.validation.Validator
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.transaction.annotation.Transactional

@SpringBootTest
@Transactional
internal class OwnerServiceTest {
    @Autowired
    lateinit var ownerService: OwnerService

    @Autowired
    lateinit var ownerRepository: OwnerRepository

    @Autowired
    lateinit var validator: Validator

    private var owner: Owner? = null

    @BeforeEach
    fun setUp() {
        ownerRepository.deleteAll()
        owner = Owner(
            id = null,
            firstName = "John",
            lastName = "Doe",
            address = "123 Main St",
            city = "Springfield",
            telephone = "123456789"
        )
    }

    @Test
    fun shouldValidateOwnerRequest_emptyDTO_failsValidation() {
        val emptyOwner = Owner()

        val violations = validator.validate<Owner?>(emptyOwner)
        violations.size shouldBe 2
        violations.map { it.propertyPath.toString() } shouldContainAll listOf("firstName", "lastName")
    }

    @Test
    fun shouldFindOwnerById() {
        val saved = ownerService.save(owner)

        val result = ownerService.findById(saved.id)

        result.shouldNotBeNull()
        result.firstName shouldBe "John"
        result.lastName shouldBe "Doe"
    }

    @Test
    fun shouldUpdateOwner() {
        val saved = ownerService.save(owner)
        saved.firstName = "Johnny"
        saved.lastName = "Smith"

        val updated = ownerService.update(saved.id, saved)
        updated.shouldNotBeNull()
        updated.firstName shouldBe "Johnny"
        updated.lastName shouldBe "Smith"
        ownerRepository.count() shouldBe 1
    }

    @Test
    fun shouldDeleteOwner() {
        val saved = ownerService.save(owner)
        val id = saved.id

        ownerService.delete(id)
        ownerService.findById(id).shouldBeNull()
        ownerRepository.count() shouldBe 0
    }

    @Test
    fun shouldCreateOwner() {
        val result = ownerService.save(owner)
        result.shouldNotBeNull()
        result.id.shouldNotBeNull()
        result.firstName shouldBe "John"
        result.lastName shouldBe "Doe"
        ownerRepository.count() shouldBe 1
    }

    @Test
    fun shouldFindAllOwners() {
        ownerService.save(owner)
        val owner2 = Owner(
            firstName = "Jane",
            lastName = "Smith",
            address = "456 Oak Ave",
            city = "Shelbyville",
            telephone = "987654321"
        )
        ownerService.save(owner2)

        val results = ownerService.findAll()

        results.shouldNotBeNull()
        results.size shouldBe 2
        results[0].firstName shouldBe "John"
        results[1].firstName shouldBe "Jane"
    }
}
